### PR TITLE
Scale up Deepseek for more layers

### PR DESCRIPTION
**Pull request by pprajapatiTT**
### What's changed
Add support for running MOE decoder layers
Create cache for tensors converted in ttnn format and load them in weight config creation
Fix metalayer mapping and padding func for MOE decoders execution
etc.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

[All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
[(Galaxy) DeepSeek tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-deepseek-tests.yaml)

### CODE REVIEW

File: models/demos/deepseek_v3/tests/test_mlp_1d.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def run_weight_conversion_test(MLPClass, hf_config, state_dict, tmp_path, refere
     # assert Path(weight_config["w3"]["input_tensor_b"]).exists()
 
     # Load and verify a weight
-    w1_ttnn = load_weight(weight_config["w1"]["input_tensor_b"], device=mesh_device)
+    w1_ttnn = ttnn.load_tensor(weight_config["w1"]["input_tensor_b"]).to(device=mesh_device)
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_mlp_1d.py]: Why do we not load the tensor directly onto the device?

Comment (barci2) [reply models/demos/deepseek_v3/tests/test_mlp_1d.py]: Also, we should definitely keep using the  `load_weight` function. Why was this changed?

File: models/demos/deepseek_v3/utils/run_config.py [hunk 2]
```diff
@@ -341,16 +335,3 @@ def _convert_run_config_to_pretty_print(run_config_item: Any, indent: int = 0) -
 def is_op_config(obj: Any) -> bool:
     """Check if the object is an op config instance."""
     return issubclass(type(obj), OpConfigBase) and is_dataclass(obj)
-
```
Comment (barci2) [inline models/demos/deepseek_v3/utils/run_config.py]: let's keep load_weight for now. I will remove it in a separate PR for full weight caching.

File: models/demos/deepseek_v3/tt/model_1d.py [hunk 3]
```diff
@@ -86,21 +112,41 @@ def convert_weights(
 
     @classmethod
     def get_meta_layer_mapping(
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/model_1d.py:99]: why are we removing all the vectorization from this function? Is there a functionality reason for this?

File: models/demos/deepseek_v3/tests/test_model.py [hunk 4]
```diff
@@ -127,11 +134,33 @@ def test_forward_pass(
     paged_input_caches, torch_page_tables = paged_caches_from_torch(input_cache, dp_factor, paged_config, user_id)
 
     # Set up model config
-    weight_config = Model1D.convert_weights(hf_config_short, [state_dict], tmp_path, mesh_device)
+    weights_type = "real_weights" if use_real_weights else "random_weights"
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_model.py]: I will soon be merging a PR with the caching functionality, so let's do away without caching for now.

File: models/demos/deepseek_v3/tests/test_model.py [hunk 5]
```diff
@@ -75,22 +77,27 @@ def test_forward_pass(
...
         reference_model = DeepseekV3ForCausalLM(hf_config_short).eval().to(torch.bfloat16)
+        logger.info("Loading real weights into reference model")
         reference_model.load_state_dict(dequantize_state_dict(state_dict, hf_config_short))
 
         torch_input = torch.randint(0, hf_config_short.vocab_size - 1, (batch_size, seq_len), dtype=torch.long)
         if mode == "prefill":
             position_ids = torch.tensor([seq_len])
         else:
             position_ids = torch.randint(0, hf_config_short.max_seq_len - 1, (batch_size,))
+            position_ids = torch.zeros(
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_model.py:97]: please add moe gate forcing to the code. The PCC issue is probably because of the discreteness of expert selection. We should be checking PCC with the same experts selected as the reference model.

File: models/demos/deepseek_v3/tests/test_mlp_1d.py (hunk 1 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_mlp_1d.py]: I am putting back load_weight and removing caching as you are already working on caching things.

File: models/demos/deepseek_v3/tt/model_1d.py (hunk 3 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/model_1d.py:99]: indexing moe state_dicts was not working out well but now with latest patch i combined state_dict creation for mlp and moe, now we can use vectorized based indexing in more elegant way.

File: models/demos/deepseek_v3/tt/model_1d.py [hunk 6]
```diff
@@ -45,33 +45,59 @@ def convert_weights(
...
+        mlp_decoder_block_state_dicts = [
             [
-                (
-                    sub_state_dict(state_dict, f"model.layers.{layer_idx}.")
-                    if layer_idx < hf_config.first_k_dense_replace
-                    else None
-                )
-                for layer_idx in range(mlp_meta_layer_indices.numel())
+                sub_state_dict(state_dict, f"model.layers.{layer_idx}.")
+                if not is_padding and -1 < layer_idx < hf_config.first_k_dense_replace
```
[the remark this answers is not in the document]
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/model_1d.py]: implemented new way to create state dicts, so this comment is not application now

File: models/demos/deepseek_v3/tests/test_model.py (hunk 4 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tests/test_model.py]: removed caching

File: models/demos/deepseek_v3/tt/model_1d.py (hunk 3 above)
Comment (pprajapatiTT) [reply models/demos/deepseek_v3/tt/model_1d.py:99]: Also, creating the mapping for meta layers was wrong, so that was fixed.
e.g. for 8 layers, we were getting [0,1,2,3],[4,5,6,7], instead of interleaved.

**Review by yieldthought — APPROVED** [review]
